python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -1187,21 +1206,30 @@ def fetch_column_metadata(
                 A set of corresponding Dagster events for dbt models, with column metadata attached,
                 yielded in the order they are emitted by dbt.
         """
-        fetch_metadata = lambda invocation, event: _fetch_column_metadata(
-            invocation, event, with_column_lineage
+        fetch_metadata = lambda invocation, ctx: _fetch_column_metadata(
+            invocation, ctx, with_column_lineage
         )
-        return self._attach_metadata(fetch_metadata)
+        return self.attach_metadata(fetch_metadata)
 
-    def _attach_metadata(
+    @public
+    @experimental
+    def attach_metadata(
         self,
-        fn: Callable[[DbtCliInvocation, DbtDagsterEventType], Optional[Dict[str, Any]]],
+        fn: Callable[
+            [
+                Union[Output, AssetMaterialization, AssetCheckResult, AssetObservation],
+                DbtAttachMetadataContext,
+            ],
+            Optional[Dict[str, Any]],
+        ],
     ) -> "DbtEventIterator[DbtDagsterEventType]":
         """Runs a threaded task to attach metadata to each event in the iterator.
 
         Args:
-            fn (Callable[[DbtCliInvocation, DbtDagsterEventType], Optional[Dict[str, Any]]]):
-                A function which takes a DbtCliInvocation and a DbtDagsterEventType and returns
-                a dictionary of metadata to attach to the event.
+            fn (Callable[[Union[Output, AssetMaterialization, AssetCheckResult, AssetObservation], DbtAttachMetadataContext], Optional[Dict[str, Any]]):
```

Comment:
Thoughts on making the event part of the context instead of its own argument? Then if someone only needs they resource props they don't need to deal with it.

Reply:
Seems reasonable